contracts/src/GaspToken.sol

```diff
@@ -1,30 +1,126 @@
+        if (uniswapPool_ == address(0)) {
+            revert ZeroUniswapPool();
+        }
 
-contract Gasp is MintableERC20 {
-    constructor() MintableERC20("GaspV2", "GASPV2") {}
-}
+        uniswapPool = uniswapPool_;
+        whitelist[uniswapPool_] = true;
+        emit UniswapPoolSet(uniswapPool_);
+    }
 
-contract DevToken is MintableERC20 {
-    constructor() MintableERC20("GaspDev", "GASPDEV") {}
-}
+    function setAllowTransfers(bool allowTransfers_) external override onlyOwner {
+        if (allowTransfers) {
+            revert TransfersAlreadyAllowed();
+        }
+
+        allowTransfers = allowTransfers_;
+        emit AllowTransfersSet(allowTransfers_);
+    }
+
+    function addToWhitelist(address account) external override onlyOwner {
+        if (account == address(0)) {
+            revert ZeroWhitelistAccount();
+        }
+        if (whitelist[account]) {
+            revert AccountAlreadyWhitelisted(account);
+        }
+
+        whitelist[account] = true;
+        emit AddedToWhitelist(account);
+    }
 
-contract ArbToken is MintableERC20 {
-    constructor() MintableERC20("CrossChain Token", "CCXT") {}
-}
\ No newline at end of file
+    function removeFromWhitelist(address account) external override onlyOwner {
+        if (account == address(0)) {
+            revert ZeroWhitelistAccount();
+        }
+        if (!whitelist[account]) {
+            revert AccountNotWhitelisted(account);
+        }
+
+        delete whitelist[account];
+        emit RemovedFromWhitelist(account);
+    }
+
+    function approve(address spender, uint256 amount)
+        public
+        override(ERC20, IERC20)
+        isWhitelisted([_msgSender(), spender], IERC20.approve.selector)
+        returns (bool)
+    {
+        return super.approve(spender, amount);
+    }
+
+    function transfer(address recipient, uint256 amount)
+        public
+        override(ERC20, IERC20)
+        isWhitelisted([_msgSender(), recipient], IERC20.transfer.selector)
+        returns (bool)
+    {
+        return super.transfer(recipient, amount);
+    }
+
+    function transferFrom(address owner, address recipient, uint256 amount)
+        public
+        override(ERC20, IERC20)
+        isWhitelisted([_msgSender(), recipient], IERC20.transferFrom.selector)
+        returns (bool)
+    {
+        return super.transferFrom(owner, recipient, amount);
+    }
+
+    function increaseAllowance(address spender, uint256 addedAmount)
+        public
+        override(ERC20, IGaspToken)
+        isWhitelisted([_msgSender(), spender], IGaspToken.increaseAllowance.selector)
+        returns (bool)
+    {
+        return super.increaseAllowance(spender, addedAmount);
+    }
+
+    function decreaseAllowance(address spender, uint256 subtractedAmount)
+        public
+        override(ERC20, IGaspToken)
+        isWhitelisted([_msgSender(), spender], IGaspToken.decreaseAllowance.selector)
```

Comment:
Why is decreaseAllowance gated by the whitelist check?
Shouldn't one be able to use it even after he removed from the whitelist - so that he can be confident to prevent misuse?

Reply:
I stick to the point here that all the whitelist functionality should be treated universally. The whiltelist will contain just a small list of trusted accounts at the initial stage which are manageable by the L1 council. So it's really infeasible and can be easily fixed by whitelisting it again if it is required in theory.